packages/contracts-bedrock/INTERFACES.md

```diff
@@ -0,0 +1,190 @@
+
+### No Interface Imports in Source Contracts
+
+Contrary to common practice, source contracts for which an interface is defined SHOULD NOT use the
+interface contract. This means:
+
+- `contract Whatever is IWhatever` is NOT allowed.
+- Source contracts should not use types or other definitions from their interfaces.
+- Contracts that build on base contracts (e.g., `contract OtherWhatever is Whatever`) should not
+  import `IWhatever` or `IOtherWhatever`.
+
+### Correct Implementation Example
+
+Instead of:
+
+```solidity
+import "./IWhatever.sol";
+
+contract Whatever is IWhatever {
+    // Implementation
+}
+```
+
+Do this:
+
+```solidity
+contract Whatever {
+    // Direct implementation without importing interface
+}
+```
+
+### Reasons for This Policy
+
+1. **Automation Potential**: We aim to auto-generate interfaces in the future. Importing interfaces
+  into source contracts would prevent this automation by creating a circular dependency.
+
+2. **ABI Compatibility**: Achieving 1:1 compatibility between interface and source contract ABI
+  becomes problematic when the source contract imports other contracts along with the interface.
+  This is due to Solidity's handling of function redefinitions. See
+  [Example of ABI Compatibility Issue](#example-of-abi-compatibility-issue) below for more context.
+
+#### Example of ABI Compatibility Issue
+
+```solidity
+contract SomeBaseContract {
+    event SomeEvent();
+}
+
+interface IWhatever {
+    event SomeEvent();
+    function someOtherFunction() external;
+}
+
+contract Whatever is IWhatever, SomeBaseContract {
+    function someOtherFunction() external {}
+}
+```
+
+In this case, Solidity will return the following compilation error:
+
+```sh
+DeclarationError: Event with same name and parameter types defined twice.
+```
+
+### Importing External Interfaces
+
+Contracts CAN import interfaces for OTHER contracts. This practice helps mitigate compilation time
+issues in older Solidity versions. As Solidity improves, we plan to phase out this exception.
+
+#### Example of Allowed Interface Usage
+
+```solidity
+import "./IOtherContract.sol";
+
+contract MyContract {
+    IOtherContract public otherContract;
+
+    constructor(address _otherContractAddress) {
+        otherContract = IOtherContract(_otherContractAddress);
+    }
+
+    // Rest of the contract
+}
+```
+
+## Creating Interfaces
+
+You have several options for creating interfaces:
+
+1. Use `cast interface`:
+
+   ```sh
+   cast interface ./path/to/contract/artifact.json
+   ```
+
+2. Use `forge inspect`:
+
+   ```sh
+   forge inspect <ContractName> abi --pretty
+   ```
+
+3. Create the interface manually:
+
+   ```solidity
+   interface IMyContract {
+       function someFunction() external;
+       function anotherFunction(uint256 _param) external returns (bool);
+       // ... other functions and events
+   }
+   ```
+
+Regardless of the method chosen, ensure that your ABIs are a 1:1 match with their source contracts.
```

Comment:
Is it worth calling out what gets lost using approaches 1 and 2? e.g. I think enum definitions are not preserved, unsure what else